src/coreclr/vm/prestub.cpp

```diff
@@ -1067,7 +1054,436 @@ PCODE MethodDesc::JitCompileCodeLocked(PrepareCodeConfig* pConfig, JitListLockEn
+
+            COR_ILMETHOD_DECODER* pILHeader = ilResolver->AllocGeneratedIL(cbCode, cbSig, maxStack);
+            BYTE* pbBuffer = (BYTE*)pILHeader->Code;
+            BYTE* pbLocalSig = (BYTE*)pILHeader->LocalVarSig;
+            _ASSERTE(cbSig == pILHeader->cbLocalVarSig);
+            sl.GenerateCode(pbBuffer, cbCode);
+            sl.GetLocalSig(pbLocalSig, cbSig);
+
+            // Store the token lookup map
+            ilResolver->SetTokenLookupMap(sl.GetTokenLookupMap());
+
+            *resolver = (DynamicResolver*)ilResolver;
+            *methodILDecoder = pILHeader;
+        }
+
+        ilResolver.SuppressRelease();
+        return true;
+    }
+}
+
+HRESULT MethodDesc::GenerateUnsafeAccessor(DynamicResolver** resolver, COR_ILMETHOD_DECODER** methodILDecoder)
+{
+    STANDARD_VM_CONTRACT;
+    _ASSERTE(resolver != NULL);
+    _ASSERTE(methodILDecoder != NULL);
+    _ASSERTE(*resolver == NULL && *methodILDecoder == NULL);
+    _ASSERTE(IsIL());
+    _ASSERTE(GetRVA() == 0);
+
+    // The UnsafeAccessorAttribute is applied to methods with an
+    // RVA of 0 (for example, C#'s extern keyword).
+    const void* data;
+    ULONG dataLen;
+    HRESULT hr = GetCustomAttribute(WellKnownAttribute::UnsafeAccessorAttribute, &data, &dataLen);
+    if (hr != S_OK)
+        return E_FAIL;
+
+    UnsafeAccessorKind kind;
+    SString name;
 
+    CustomAttributeParser ca(data, dataLen);
+    if (!TryParseUnsafeAccessorAttribute(this, ca, kind, name))
+        return COR_E_BADIMAGEFORMAT;
+
+    GenerationContext context{ kind, this };
+
+    // Parse the signature to determine the type to use:
+    //  * Constructor access - examine the return type
+    //  * Instance member access - examine type of first parameter
+    //  * Static member access - examine type of first parameter
+    TypeHandle retType;
+    TypeHandle firstArgType;
+    retType = context.DeclarationSig.GetRetTypeHandleThrowing();
+    UINT argCount = context.DeclarationSig.NumFixedArgs();
+    if (argCount > 0)
+    {
+        context.DeclarationSig.NextArg();
+        firstArgType = context.DeclarationSig.GetLastTypeHandleThrowing();
+    }
+
+    // Using the kind type, perform the following:
+    //  1) Validate the basic type information from the signature.
+    //  2) Resolve the name to the appropriate member.
+    //  3) Generate the IL for the accessor.
+    switch (context.Kind)
+    {
+    case UnsafeAccessorKind::Constructor:
+        // A return type is required for a constructor, otherwise
+        // we don't know the type to construct.
+        // The name is defined by the runtime and should be empty.
+        if (retType.IsNull() || !name.IsEmpty())
+        {
+            return COR_E_BADIMAGEFORMAT;
+        }
+
+        context.TargetType = retType;
+        if (!TrySetTargetMethodCtor(context)
+            || !TryGenerateAccessor(context, resolver, methodILDecoder))
+        {
+            return COR_E_MISSINGMETHOD;
```

Comment:
Would it be better to throw exception immediately right here instead of returning HRESULT?

Reply:
Not in this case. This API is called in two distinct places. The first is during normal JIT of a method, that is `UnsafeJitFunction`. The second is in `CEEInfo::getMethodInfo()` which is commonly used for inline determination. For making the inline check we can check the `HRESULT` and follow the API by returning `false` where as if we throw then the API is more muddled with "throw" or "false" to indicate failure.

My personal preference is that boundaries across binaries or at the subsystem API level follow the APIs natural expression of failure. In this case the `HRESULT` provides that and allows the decision to be made at the correct location.

Reply:
JIT/EE interface error handling is based on exception handling. We are not going to change that. It would be a regression in user facing diagnostics experience and it would make the runtime and JIT code more complicated for no good reason.

It is fine to throw exception from `getMethodInfo` JIT/EE interface method for invalid programs. The inliner is going to deal with it correctly.

`bool` return value from `getMethodInfo` is meant to be used for situations where it is valid for the method to not have IL. We do not want to throw and catch exceptions for valid programs. But it is fine to throw and catch exceptions during JITing for invalid programs. It happens quite a bit in practice actually, e.g. for assemblies with missing dependencies.

Reply:
> JIT/EE interface error handling is based on exception handling. We are not going to change that. 

Agree, I'm not advocating for that. However, I'm also not likely to implement something that ignores an obvious mechanism (that is, return `false`) to indicate something isn't possible.

> bool return value from getMethodInfo is meant to be used for situations where it is valid for the method to not have IL. 

I'm not convince that argument works for all cases. Consider the case where a branch doesn't need to be taken. Throwing an exception at this point is a pessimisation as it introduces cost that is unnecessary because the method can't be inlined, but that is potentially okay for a user because the path isn't taken.

Unless there is a performance or architectual issue I prefer to use `HRESULT`s in this path.

Reply:
> Consider the case where a branch doesn't need to be taken. Throwing an exception at this point is a pessimisation as it introduces cost that is unnecessary

It is perfectly fine to pessimize this case. We are not optimizing for invalid programs, even when the invalid part happens to be unreachable. We are not trying to avoid this cost for dynamically unreachable invalid parts in hundreds of existing situations. Why should this one be special?

For me, the issue is consistency and diagnostic experience. For example, you can throw exception with better error message that includes detail about what was not found or what was otherwise wrong.

Reply:
`getMethodInfo` throws for invalid programs today (e.g. if the metadata is invalid). It returns false only for cases where the method is expected to not have the IL.

Reply:
Done.